File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/read/PebbleReadPetroglyph.kt
package app.pbbls.android.features.path.read

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import app.pbbls.android.features.path.models.EmotionPalette
import app.pbbls.android.features.path.models.Valence
import app.pbbls.android.features.path.render.PebbleOutlineBackdrop
import app.pbbls.android.features.path.render.PebbleOutlineGeometry
import app.pbbls.android.features.path.render.PebbleStaticRender
import app.pbbls.android.theme.PebblesTheme

/**
 * The read-view "Petroglyph" (issue #599): the outline silhouette backfill with
 * the server-composed render (outline + glyph) traced on top, down-scaled per
 * [PebbleOutlineGeometry] to sit inside the silhouette. Same composite the Path
 * rows draw via `PebbleThumbnail`, but coloured through [petroglyphColors] so
 * the read page follows #599's per-size, per-theme table instead of the shared
 * theme-neutral `pebbleFrameColors`.
 *
 * Sized by the caller's [modifier]: the heading treatment gives it a large
 * square slot, the snap overlay a smaller one. A missing [palette] (cache cold,
 * unknown or absent emotion) falls back to the brand accent, mirroring
 * `PebbleThumbnail`. [palette] and the valence data arrive as parameters so
 * screenshot previews drive it without a live client.
 */
@Composable
fun PebbleReadPetroglyph(
    renderSvg: String?,
    valence: Valence,
    palette: EmotionPalette?,
    modifier: Modifier = Modifier,
) {
    val accentHex = PebblesTheme.colors.accent.primaryHex
    val sizeGroup = valence.sizeGroup
    val colors =
        palette?.let { petroglyphColors(it, sizeGroup, isSystemInDarkTheme()) }
            ?: PetroglyphColors(strokeHex = accentHex, fillHex = accentHex, fillOpacity = 1f)

    Box(modifier, contentAlignment = Alignment.Center) {
        Box(
            modifier = Modifier.aspectRatio(PebbleOutlineGeometry.aspectRatio(sizeGroup)),
            contentAlignment = Alignment.Center,
        ) {
            PebbleOutlineBackdrop(
                sizeGroup = sizeGroup,
                polarity = valence.polarity,
                fillHex = colors.fillHex,
                fillOpacity = colors.fillOpacity,
                modifier = Modifier.fillMaxSize(),
            )
            if (renderSvg != null) {
                PebbleStaticRender(
                    svg = renderSvg,
                    strokeHex = colors.strokeHex,
                    modifier =
                        Modifier
                            .fillMaxSize()
                            .scale(PebbleOutlineGeometry.pebbleScale(sizeGroup)),
                )
            }
        }
    }
}

File: apps/ios/Pebbles/Features/Path/Read/PebbleReadPetroglyph.swift
import SwiftUI

/// The read-view "Petroglyph" (issue #599): the framed pebble — outline
/// silhouette backfill with the composed render (outline + glyph) traced on
/// top — shown either as the page heading (no snap) or overlapping the snap's
/// top-right corner (`PebbleSnapFrame`).
///
/// Reuses `PebbleAnimatedRenderView`, which already composites the backdrop and
/// the pebble with the read view's native draw-on animation; only the coloring
/// changes here — from the theme-neutral `pebbleFrameColors` used by the Path
/// rows to #599's per-size, per-scheme `petroglyphColors` table.
///
/// Sized by the caller: the heading treatment gives it a large square slot, the
/// snap overlay a smaller one (the render fits its own outline aspect inside
/// whatever frame it gets). A missing `palette` (cache cold, unknown or absent
/// emotion) falls back to the brand accent, mirroring `PathPebbleRow`. `palette`
/// and the valence data arrive as parameters so previews drive it without a
/// live client.
struct PebbleReadPetroglyph: View {
    let renderSvg: String?
    let renderVersion: String?
    let valence: Valence
    let palette: EmotionPalette?

    @Environment(\.colorScheme) private var colorScheme

    var body: some View {
        if let renderSvg {
            let colors = resolvedColors
            PebbleAnimatedRenderView(
                svg: renderSvg,
                strokeColor: Color(hex: colors.strokeHex) ?? Color.accent.primary,
                strokeColorHex: colors.strokeHex,
                fillHex: colors.fillHex,
                fillOpacity: colors.fillOpacity,
                size: valence.sizeGroup,
                polarity: valence.polarity,
                renderVersion: renderVersion
            )
        }
    }

    /// #599 colors for the current size + scheme, or the brand accent when the
    /// palette is unavailable (opaque accent for both stroke and fill).
    private var resolvedColors: PetroglyphColors {
        guard let palette else {
            let accentHex = Color.accent.primaryHex
            return PetroglyphColors(strokeHex: accentHex, fillHex: accentHex, fillOpacity: 1)
        }
        return palette.petroglyphColors(forSize: valence.sizeGroup, scheme: colorScheme)
    }
}
